glusterblock/glusterblock-provisioner.go

```diff
@@ -0,0 +1,355 @@
+
+	secretMap, err := GetSecretForPV(namespace, secretName, "glusterblock", kubeClient)
+	if err != nil {
+		glog.Errorf("failed to get secret %s/%s: %v", namespace, secretName, err)
+		return "", fmt.Errorf("failed to get secret %s/%s: %v", namespace, secretName, err)
+	}
+	if len(secretMap) == 0 {
+		return "", fmt.Errorf("empty secret map")
+	}
+	secret := ""
+	for k, v := range secretMap {
+		if k == secretKeyName {
+			return v, nil
+		}
+		secret = v
+	}
+	// If not found, the last secret in the map wins as done before
+	return secret, nil
+}
+
+
+// GetSecretForPV locates secret by name and namespace, verifies the secret type, and returns secret map
+func GetSecretForPV(secretNamespace, secretName, volumePluginName string, kubeClient clientset.Interface) (map[string]string, error) {
+        secret := make(map[string]string)
+        if kubeClient == nil {
+                return secret, fmt.Errorf("Cannot get kube client")
+        }
+        secrets, err := kubeClient.Core().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
+        if err != nil {
+                return secret, err
+        }
+        if secrets.Type != v1.SecretType(volumePluginName) {
+                return secret, fmt.Errorf("Cannot get secret of type %s", volumePluginName)
+        }
+        for name, data := range secrets.Data {
+                secret[name] = string(data)
+        }
+        return secret, nil
+}
+
+var (
+	master     = flag.String("master", "", "Master URL")
+	kubeconfig = flag.String("kubeconfig", "", "Absolute path to the kubeconfig")
+)
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	// Create an InClusterConfig and use it to create a client for the controller
+	// to use to communicate with Kubernetes
+
+	var config *rest.Config
+	var err error
+	if *master != "" || *kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		glog.Fatalf("Failed to create config: %v", err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Failed to create client: %v", err)
+	}
+
+	// The controller needs to know what the server version is because out-of-tree
+	// provisioners aren't officially supported until 1.5
+	serverVersion, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		glog.Fatalf("Error getting server version: %v", err)
+	}
+
+	// Create the provisioner: it implements the Provisioner interface expected by
+	// the controller
+	glusterBlockProvisioner := NewglusterBlockProvisioner(clientset)
+
+	// Start the provision controller which will dynamically provision glusterblock
+	// PVs
+	pc := controller.NewProvisionController(clientset, resyncPeriod, provisionerName, glusterBlockProvisioner, serverVersion.GitVersion, exponentialBackOffOnError, failedRetryThreshold)
```

Comment:
the latest version of library requires 4 new arguments at the end https://github.com/kubernetes-incubator/external-storage/blob/master/nfs/cmd/nfs-provisioner/main.go#L120 

usually I just use default like this and I have to import the leaderelection package https://github.com/kubernetes-incubator/external-storage/blob/master/nfs/cmd/nfs-provisioner/main.go#L53